### ISSUE TITLE
ZIO Test: Refactor FailureDetails

### ISSUE THREAD

**Issue by jdegoes** [zio/zio#1332]
### 1. Rename `FailureDetails` => `TestFailure`

Every `FailureDetails` represents the failure of a single test, so `TestFailure` is more descriptive and more consistent with other constructs inside the library.

### 2. Allow stack of predicate values

When a `FailureDetails` represents the failure of a predicate, currently there are two bits of information stored:

```scala
final case class Predicate(fragment: PredicateValue, whole: PredicateValue) extends FailureDetails
```

The `fragment` `PredicateValue`, and the `whole`. The first is the "inner most" failure, while the second is the "outer most" failure. For example, in a composite predicate like `isSome(isGreaterThan(4))`, if the value is a `Some`, but is 3, then there would be two failures:

 - `Some(3) did not satisfy isSome(isGreaterThan(4))`
 - `3 did not satisfy isGreaterThan(4)`

Instead of restricting to just 2, we should have a stack, so we can have 1 or more `PredicateValue`, potentially allowing a whole stack, each level "drilling down" into the failure.

This would also allow us to have 1 failure, rather than 2. In some cases, like `equals(32)`, there is just one failure, not two, so rather than de-duplicating that failure (as per work of @adamgfraser), we should just not create it.

The new representation would be:

```scala
final case class Predicate(values: ::[PredicateValue]) extends FailureDetails
```

where `::` is the "cons" constructor of `List`, which guarantees a head and a tail.

This change will necessitate modifying `Predicate.scala`, to "push" a new `PredicateValue` onto the stack in every combinator (for example, `isSome(pred)` is a combinator, because it is a function that takes a predicate and returns a predicate). 

I can assist with this if it is unclear.

**Comment by mgibowski** [comment]
I would like to take this.
I took a look at code and for now it seems clear.

**Comment by jdegoes** [comment]
@mgibowski Awesome. Thank you! 🙏

**Comment by mgibowski** [comment]
Sorry for taking time. I was expecting to find free time for coding more easily. Finally I am working on this.

@jdegoes  I would like to clarify something.

> Instead of restricting to just 2, we should have a stack, so we can have 1 or more PredicateValue, potentially allowing a whole stack, each level "drilling down" into the failure.

In case when the stack has more than one failure - we should report all of them rather then only  the first inner one, right?

For example what we want is:
<img width="786" alt="expected" src="https://user-images.githubusercontent.com/617315/63167405-1e890880-c029-11e9-9482-1070aff9f11d.png">

Rather than (currently):
<img width="814" alt="now" src="https://user-images.githubusercontent.com/617315/63167416-247ee980-c029-11e9-81c8-8aa2d29677ee.png">

Correct?

Side note: at the moment the meaning of `isGreaterThan` and `isLessThan` seems inverted. I will fix that as part of the PR.

**Comment by adamgfraser** [comment]
@mgibowski We want to report all the inner failures when there are multiple nested in failures. For example, if we had a predicate like

```scala
isRight(isSome(isGreaterThan(4)))
```

We would expect to get three levels of error reporting like so:

```scala
Right(Some(3)) did not satisfy isRight(isSome(isGreaterThan(4)))
Some(3) did not satisfy isSome(isGreaterThan(4))
3 did not satisfy isGreaterThan(4)
```

In your example with the `&&` clause we are fine with reporting just the failure. Otherwise we would need a tree instead of a nonempty list to capture the different failures at each level and their respective nested failures, which I think is more complexity than we need here.

You are 100% correct about `isLessThan` and `isGreaterThan` methods. Ironically we have test cases for those but they are testing exactly the opposite of the correct behavior.

**Comment by mgibowski** [comment]
Thanks @adamgfraser! 🙏

**Comment by jdegoes** [comment]
What @adamgfraser said. 😄

**Comment by mgibowski** [comment]
@adamgfraser @jdegoes 
I started having doubts if such reporting is really more user-friendly.

For example here we have very exact information about what failed:
<img width="731" alt="Zrzut ekranu 2019-08-16 o 15 12 51" src="https://user-images.githubusercontent.com/617315/63174259-82b3c880-c039-11e9-9e6c-0b03cbc66dc0.png">
_(I still use the inverted `isGreaterThan`, please ignore that)_

If we transform this into:
<img width="719" alt="Zrzut ekranu 2019-08-16 o 15 13 07" src="https://user-images.githubusercontent.com/617315/63174325-a37c1e00-c039-11e9-9a18-3eb57ed5aa51.png">

then imho it doesn't help very much. `isSome` doesn't fail because the value is not `Some`, it fails only because of the nested predicate. If we report also all the intermediary successful (in their context) predicates, then I see this as introducing more noise to the reader.

As a user in case of such a failure I only care about `isGreaterThen(2)` part.

What is the value of presenting all the intermediary predicates that wouldn't fail if they didn't contain a nested failure?

**Comment by adamgfraser** [comment]
@mgibowski I think there may be a distinction between the failure information we want to print to the user in the default test reporter and the failure representation we use internally.

I can see your argument about the intermediate predicates being unnecessary in this case but I wonder if that is just because this is an example with a couple of levels of nesting but not that many. You can think of the failure details as almost a "stack trace" of the failure and like a normal stack trace you're right that a lot of times we only care about the "top" (my program blew up) and the "bottom" (there was a `NoSuchElementException` at line 25) but sometimes the intermediate representation is helpful. If we keep a stack of the failures then the test reporter can decide what combination of failure details to print, versus if we just keep the top and the bottom the test reporter doesn't have that option anymore.

The other aspect that motivated this is that a lot of times the stack representation may actually let us create fewer failure details. Right now for every failure we're recording a "whole" and a "fragment", but for simple predicates that don't have nesting the whole is identical to the fragment, so we're essentially recording the same information twice and then the test reporter has to have special logic to deduplicate those so it doesn't print the same thing twice, whereas if we used a stack we would only ever push one failure to the stack.

Does that make sense?

**Comment by mgibowski** [comment]
@adamgfraser thanks for your answer!

> The other aspect that motivated this is that a lot of times the stack representation may actually let us create fewer failure details. Right now for every failure we're recording a "whole" and a "fragment"

Replacing the `fragment` and `whole` with a stack for the reason above makes complete sense to me.

However, I am still not convinced that storing intermediary predicates on the stack without presenting them to the user is useful. No use case when that would be helpful comes to my mind. Adding more levels of nesting doesn't change the point imho.

I think the case of stack traces is a bit different. You may have code in your module calling code from another module, then calling some library and in the end terminating with some exception. In such case it is really useful to see the whole stack trace and be able to investigate which exactly line of your main module did invoke which line of another module, which in turn called some library and resulted with the error...

But in our case it is different. We only have a nested value constructed by a user in his test:
`not(isRight(isSome(isRight(isGreaterThan(4)))))`.

I currently have working code basically ready that replaces old `fragment` and `whole` with a stack without accumulating intermediary predicates. I can try experimenting with an implementation of the full stack, but skipping intermediary predicates for default reporting. Nevertheless it would be good to understand, why we need it :)

**Comment by adamgfraser** [comment]
@mgibowski Totally makes sense. I'll defer to @jdegoes on his vision for this.

**Comment by jdegoes** [comment]
@mgibowski If you see something like `3 did not satisfy isGreaterThan(4)`, then in a _complex_ predicate, with multiple `isGreaterThan(x)`, you will not know _which_ one failed without context—the stack is one way to provide context. Indeed, one could use the stack when rendering to decide which of the `isGreaterThan(x)` actually failed (even though that's not currently done).

Moreover there's a deeper and stronger reason to eliminate the `whole` and `fragment`: it forces the top-level assertion to keep track of a `PredicateValue` failure; this creates an asymmetry that necessitates the "de-duplication" in the current reporter, and will have more complex consequences when assertions can be composed.

Basically these pains are telling us the special-case logic has outlived its usefulness. Eliminating the special-case logic, in favor of a composable stack, will not only eliminate de-duplication, but improve composition of assertions and provide more context for failures.

**Comment by adamgfraser** [comment]
@mgibowski Haven't heard back on this in a while so I am going to start working on this.

**Comment by mgibowski** [comment]
@adamgfraser thanks! sorry, I don't have the notifications on.
My work on this got interrupted with a family travel, I am glad you took it over. I will be happy to help with forthcoming things.
